pkg/ipamd/ipamd.go

```diff
@@ -1186,6 +1374,79 @@ func (c *IPAMContext) verifyAndAddIPsToDatastore(eni string, attachedENIIPs []*e
 	return seenIPs
 }
 
+// verifyAndAddPrefixesToDatastore updates the datastore with the known Prefixes. Prefixes who are out of cooldown gets added
+// back to the datastore after being verified against EC2.
+func (c *IPAMContext) verifyAndAddPrefixesToDatastore(eni string, attachedENIIPs []*ec2.Ipv4PrefixSpecification, needEC2Reconcile bool) map[string]bool {
+	var ec2VerifiedAddresses []*ec2.Ipv4PrefixSpecification
+	seenIPs := make(map[string]bool)
+	for _, privateIPv4 := range attachedENIIPs {
+		strPrivateIPv4 := aws.StringValue(privateIPv4.Ipv4Prefix)
+		log.Debugf("Check in coolddown Found prefix %s", strPrivateIPv4)
+
+		// Check if this Prefix was recently freed
+		_, ipv4CidrPtr, err := net.ParseCIDR(strPrivateIPv4)
+		if err != nil {
+			log.Debugf("Failed to parse so continuing with next prefix")
+			continue
+		}
+		found, recentlyFreed := c.reconcileCooldownCache.RecentlyFreed(strPrivateIPv4)
+		if found {
+			log.Debugf("found in cooldown")
+			if recentlyFreed {
+				log.Debugf("Reconcile skipping IP %s on ENI %s because it was recently unassigned from the ENI.", strPrivateIPv4, eni)
+				continue
+			} else {
+				if needEC2Reconcile {
+					// IMDS data might be stale
+					log.Debugf("This IP was recently freed, but is now out of cooldown. We need to verify with EC2 control plane.")
+					// Only call EC2 once for this ENI
+					if ec2VerifiedAddresses == nil {
+						var err error
+						// Call EC2 to verify IPs on this ENI
+						ec2VerifiedAddresses, err = c.awsClient.GetIPv4PrefixesFromEC2(eni)
+						if err != nil {
+							log.Errorf("Failed to fetch ENI IP addresses from EC2! %v", err)
+							// Do not delete this IP from the datastore or cooldown until we have confirmed with EC2
+							seenIPs[strPrivateIPv4] = true
+							continue
+						}
+					}
+					// Verify that the IP really belongs to this ENI
+					isReallyAttachedToENI := false
+					for _, ec2Addr := range ec2VerifiedAddresses {
+						if strPrivateIPv4 == aws.StringValue(ec2Addr.Ipv4Prefix) {
+							isReallyAttachedToENI = true
+							log.Debugf("Verified that IP %s is attached to ENI %s", strPrivateIPv4, eni)
+							break
+						}
+					}
+					if !isReallyAttachedToENI {
+						log.Warnf("Skipping IP %s on ENI %s because it does not belong to this ENI!", strPrivateIPv4, eni)
+						continue
+					}
+				}
+				// The IP can be removed from the cooldown cache
+				// TODO: Here we could check if the IP is still used by a pod stuck in Terminating state. (Issue #1091)
+				c.reconcileCooldownCache.Remove(strPrivateIPv4)
+			}
+		}
+
+		err = c.dataStore.AddIPv4CidrToStore(eni, *ipv4CidrPtr, true)
```

Comment:
so no matter whether the CIDR is in really attached we added it to the store, is the logic correct?

Reply:
i see, the continue breaks the for loop in 5 levels above..

Reply:
Yeah, the 2 loop comment should fix this too :D